opal/mca/memory/patcher/memory_patcher_component.c

```diff
@@ -404,6 +407,68 @@ static size_t memory_patcher_get_shm_seg_size (const void *shmaddr)
     return seg_size;
 }
 
+static size_t get_shm_size(int shmid)
+{
+    struct shmid_ds ds;
+    int ret;
+
+    ret = shmctl(shmid, IPC_STAT, &ds);
+    if (ret < 0) {
+        return 0;
+    }
+
+    return ds.shm_segsz;
+}
+#endif
+
+#if HAS_SHMAT && defined(__linux__)
+static void *(*original_shmat)(int shmid, const void *shmaddr, int shmflg);
+
+static void *_intercept_shmat(int shmid, const void *shmaddr, int shmflg)
+{
+    void *result = 0;
+
+    size_t size = get_shm_size(shmid);
+
+    if ((shmflg & SHM_REMAP) && (shmaddr != NULL)) {
+// I don't really know what REMAP combined with SHM_RND does, so I'll just
+// guess it remaps all the way down to the lower attach_addr, and all the
+// way up to the original shmaddr+size
+        uintptr_t attach_addr = (uintptr_t)shmaddr;
+        
+        if (shmflg & SHM_RND) {
+            attach_addr -= attach_addr % SHMLBA;
+            size += attach_addr % SHMLBA;
```

Comment:
isn't  `attach_addr % SHMLBA` equal to 0 now, after subtracting the remainder?

Reply:
yeah, that's definitely not what I meant to do